src/main/java/com/iota/iri/service/milestone/impl/MilestoneServiceImpl.java

```diff
@@ -312,5 +367,41 @@ private boolean isMilestoneBundleStructureValid(List<TransactionViewModel> bundl
                 .allMatch(branchTransactionHash -> branchTransactionHash.equals(headTransactionHash));
     }
 
+    /**
+     * This method does the same as {@link #resetCorruptedMilestone(int)} but additionally receives a set of {@code
+     * processedTransactions} that will allow us to not process the same transactions over and over again while
+     * resetting additional milestones in recursive calls.<br />
+     * <br />
+     * It first checks if the desired {@code milestoneIndex} is reachable by this node and then triggers the reset
+     * by:<br />
+     * <br />
+     * 1. resetting the ledger state if it addresses a milestone before the current latest solid milestone<br />
+     * 2. resetting the {@code milestoneIndex} of all transactions that were confirmed by the current milestone<br />
+     * 3. deleting the corresponding {@link StateDiff} entry from the database<br />
+     *
+     * @param index milestone index that shall
+     * @param processedTransactions a set of transactions that have been processed already
+     * @throws MilestoneException if anything goes wrong while resetting the corrupted milestone
+     */
+    private void resetCorruptedMilestone(int index, Set<Hash> processedTransactions) throws MilestoneException {
+        if(index <= snapshotProvider.getInitialSnapshot().getIndex()) {
+            return;
+        }
+         log.info("resetting corrupted milestone #" + index);
+         try {
+            MilestoneViewModel milestoneToRepair = MilestoneViewModel.get(tangle, index);
+            if(milestoneToRepair != null) {
+                if(milestoneToRepair.index() <= snapshotProvider.getLatestSnapshot().getIndex()) {
+                    snapshotService.rollBackMilestones(snapshotProvider.getLatestSnapshot(), milestoneToRepair.index());
+                }
+                 updateMilestoneIndexOfMilestoneTransactions(milestoneToRepair.getHash(), milestoneToRepair.index(), 0,
+                        processedTransactions);
+                tangle.delete(StateDiff.class, milestoneToRepair.getHash());
```

Comment:
what happens if we fail before this line is processed? what will happen when IRI is reset?
(i.e. non-atomic change)

Reply:
It will lead to an inconsistency which can also happen during the application of milestones (in contrast to reverting them).

At the moment this will be fixed by the repair mechanism. The repair will just be triggered again if the milestone is still inconsistent (after a IRI restart).